### PR TITLE
Add AOI validations to geoprocessing API

### PR DESCRIPTION
**Pull request by kellyi**
## Overview

This PR adds some validations on the size & structure of the AOI for the analyze & climate endpoints. I put everything complex into a validations.py file -- including most of the RWD validations from #2386 as well as the functions to generate the error messages.

For the AOI endpoints we now verify that the shape does not exceed 75k sq km and that it passes the `.valid` check from django gis/GEOSGeometry. If it fails the latter, we send back an error message describing what failed.

Note: the second commit here includes the validation to check whether a shape is within the continental US boundaries.

Connects #2340 
Connects #2342 
Connects #2343 

### Notes

I noted this on #2342 but the `.valid` method seems to catch more than just self-intersecting shapes. If we decide we want to filter out only the self-intersecting ones but accept shapes which otherwise fail the `.valid` check, we can update the method just to check on what the error message would be & return a bool from there.

## Testing Instructions
- get this branch, then bundle, and visit the api docs with your api key at the ready
- check the RWD inputs to verify that they still work as described in #2386 
- for the AOI inputs, test out the following three shapes:

1. Invalid because it's 82k sq km:

```json
{
        "type": "MultiPolygon",
        "coordinates": [[
          [
            [
              -83.0950927734375,
              34.99850370014629
            ],
            [
              -83.36975097656249,
              34.728069689872285
            ],
            [
              -81.090087890625,
              32.18026188320708
            ],
            [
              -80.61767578124999,
              32.20350534542368
            ],
            [
              -78.5577392578125,
              33.8247936182649
            ],
            [
              -79.7003173828125,
              34.82282272723702
            ],
            [
              -80.771484375,
              34.80929324176267
            ],
            [
              -81.06262207031249,
              35.16482750605027
            ],
            [
              -82.2601318359375,
              35.21869749632885
            ],
            [
              -83.0950927734375,
              34.99850370014629
            ]
          ]
        ]]
      }
```

2. invalid because it self-intersects:

```json
{
        "type": "MultiPolygon",
        "coordinates": [[
          [
            [
              -75.4376220703125,
              40.14109012528468
            ],
            [
              -75.5364990234375,
              39.609920257000795
            ],
            [
              -74.739990234375,
              40.16208338164617
            ],
            [
              -74.8828125,
              39.69450749856091
            ],
            [
              -75.4376220703125,
              40.14109012528468
            ]
          ]
        ]]
      }
```

3. valid, about 66k sq km:

```json
{
        "type": "MultiPolygon",
        "coordinates": [[
          [
            [
              -80.628662109375,
              40.622291783092706
            ],
            [
              -81.002197265625,
              39.54641191968671
            ],
            [
              -82.529296875,
              38.436379603
            ],
            [
              -82.672119140625,
              38.151837403006766
            ],
            [
              -81.650390625,
              37.07271048132943
            ],
            [
              -80.167236328125,
              37.61423141542417
            ],
            [
              -79.617919921875,
              38.436379603
            ],
            [
              -78.233642578125,
              39.444677580473424
            ],
            [
              -78.75,
              39.631076770083666
            ],
            [
              -79.46411132812499,
              39.2492708462234
            ],
            [
              -79.4970703125,
              39.715638134796336
            ],
            [
              -80.518798828125,
              39.715638134796336
            ],
            [
              -80.5078125,
              40.63896734381723
            ],
            [
              -80.628662109375,
              40.622291783092706
            ]
          ]
        ]]
      }
```

### CODE REVIEW

**Comment by kellyi** [comment]
I pushed another commit to validate that the aoi shapes are within CONUS boundaries, too. This AOI should work:

```json
{
        "type": "MultiPolygon",
        "coordinates": [[
          [
            [
              -82.5732421875,
              28.275358281817105
            ],
            [
              -82.4853515625,
              27.42053815128712
            ],
            [
              -81.925048828125,
              27.488781168937997
            ],
            [
              -82.30957031249999,
              28.835049972635176
            ],
            [
              -82.63916015625,
              28.76765910569123
            ],
            [
              -82.5732421875,
              28.275358281817105
            ]
          ]
        ]]
      }
```

and this aoi should show an error reading

```json
[
  "Area of interest boundaries must be within the boundaries of the continental United States"
]
```

```json
{
        "type": "MultiPolygon",
        "coordinates": [[
          [
            [
              -125.1123046875,
              49.18170338770663
            ],
            [
              -123.96972656249999,
              46.649436163350245
            ],
            [
              -122.08007812499999,
              46.9502622421856
            ],
            [
              -122.08007812499999,
              50.0923932109388
            ],
            [
              -125.1123046875,
              49.18170338770663
            ]
          ]
        ]]
      }
```

File: src/mmw/apps/geoprocessing_api/views.py [hunk 1]
```diff
@@ -800,6 +835,18 @@ def start_analyze_pointsource(request, format=None):
     wkaoi = request.query_params.get('wkaoi', None)
     area_of_interest = load_area_of_interest(request.data, wkaoi)
 
+    if not validate_analyze_aoi_size_below_max_area(area_of_interest):
```
Comment (kellyi) [inline src/mmw/apps/geoprocessing_api/views.py]: It occurred to me that we could just add a single `validate_aoi` function to group all these together in the `validations.py` file then run it there.

Comment (kellyi) [reply src/mmw/apps/geoprocessing_api/views.py]: Done in 45de649

**Comment by arottersman** [comment]
Taking a look now

File: src/mmw/apps/geoprocessing_api/validation.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    return ('Invalid optional `snappingOn` parameter value `{}`. Must be '
+            '`true` or `false`.'.format(snapping))
+
+
+def create_invalid_rwd_simplify_param_type_error_msg(simplify):
+    return ('Invalid optional `simplify` parameter value: `{}`. Must be a '
+            'number.'.format(simplify))
+
+
+def validate_rwd_simplify_param_type(simplify):
```
Comment (arottersman) [inline src/mmw/apps/geoprocessing_api/validation.py]: Consider creating a single function `validate_rwd_request_payload` that checks all parts of the body for errors, and raises a validation error with the list of things that are wrong.

File: src/mmw/apps/geoprocessing_api/validation.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+def validate_analyze_aoi_size_below_max_area(aoi):
+    return get_aoi_sq_km(aoi) < settings.MMW_MAX_AREA
+
+
+def create_invalid_shape_error_msg(aoi):
+    return ('Area of interest is invalid: {}'.format(
+            GEOSGeometry(aoi, 4326).valid_reason))
+
+
+def validate_aoi_does_not_self_intersect(aoi):
```
Comment (arottersman) [inline src/mmw/apps/geoprocessing_api/validation.py]: Consider renaming these to some other verb like `check` to reserve `validate` only for functions in this file that raise proper `ValidationErrors`. Alternatively, move the `ValidationError` in `validate_aoi` here.

**Comment by kellyi** [comment]
Thanks for the feedback! I adjusted the validation method names & structure to simplify them in 
 ddb021b1815e9869747714f867de9f9f5ef9074e

Only methods of a type like `Maybe(ValidationError)` keep the `validate_` name prefix; actual checks go in `check_` and we now only import the `validate_rwd` and `validate_aoi` methods into `views.py`.

**Comment by kellyi** [comment]
Thanks! I squashed this & rewrote the commit message and tested it out again locally. Still works, so going to merge it in once the tests pass.

**Pull request merged**